agent/consul/stream/event_publisher.go

```diff
@@ -0,0 +1,305 @@
+package stream
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// EventPublisher receives changes events from Publish, and sends them to all
+// registered subscribers.
+type EventPublisher struct {
+	// topicBufferSize controls how many trailing events we keep in memory for
+	// each topic to avoid needing to snapshot again for re-connecting clients
+	// that may have missed some events. It may be zero for no buffering (the most
+	// recent event is always kept though). TODO
+	topicBufferSize int
+
+	// snapCacheTTL controls how long we keep snapshots in our cache before
+	// allowing them to be garbage collected and a new one made for subsequent
+	// requests for that topic and key. In general this should be pretty short to
+	// keep memory overhead of duplicated event data low - snapshots are typically
+	// not that expensive, but having a cache for a few seconds can help
+	// de-duplicate building the same snapshot over and over again when a
+	// thundering herd of watchers all subscribe to the same topic within a few
+	// seconds.
+	snapCacheTTL time.Duration
+
+	// This lock protects the topicBuffers, and snapCache
+	lock sync.RWMutex
+
+	// topicBuffers stores the head of the linked-list buffer to publish events to
+	// for a topic.
+	topicBuffers map[Topic]*eventBuffer
+
+	// snapCache if a cache of EventSnapshots indexed by topic and key.
+	// TODO: new struct for snapCache and snapFns and snapCacheTTL
```

Comment:
Is this your TODO Daniel? If yes it might help to scope it like `TODO(streaming)` so we can review them all at the end and see if it's important or can be removed?